In [2]:
pip install yfinance

     -------------------------------------- 119.4/119.4 KB 3.5 MB/s eta 0:00:00
     ---------------------------------------- 64.8/64.8 KB ? eta 0:00:00
     -------------------------------------- 187.3/187.3 KB 5.7 MB/s eta 0:00:00
     ------------------------------------- 949.2/949.2 KB 10.0 MB/s eta 0:00:00
  Installing build dependencies: started
  Installing build dependencies: finished with status 'done'
  Getting requirements to build wheel: started
  Getting requirements to build wheel: finished with status 'done'
  Preparing metadata (pyproject.toml): started
  Preparing metadata (pyproject.toml): finished with status 'done'
     ---------------------------------------- 1.6/1.6 MB 12.8 MB/s eta 0:00:00
     ------------------------------------- 176.8/176.8 KB 10.4 MB/s eta 0:00:00
     ------------------------------------- 435.3/435.3 KB 26.6 MB/s eta 0:00:00
  Preparing metadata (setup.py): started
  Preparing metadata (setup.py): finished with status 'done'
     -----------

You should consider upgrading via the 'c:\Users\user\AppData\Local\Programs\Python\Python39\python.exe -m pip install --upgrade pip' command.


In [1]:
import yfinance as yf

yf.__version__

'0.2.65'

In [2]:
df = yf.download(['MSFT','AAPL','GOOG'], period='1mo')
df

C:\Users\user\AppData\Local\Temp\ipykernel_12756\2261837217.py:1: FutureWarning: YF.download() has changed argument auto_adjust default to True
  df = yf.download(['MSFT','AAPL','GOOG'], period='1mo')
[*********************100%***********************]  3 of 3 completed


Price            Close                                High              \
Ticker            AAPL        GOOG        MSFT        AAPL        GOOG   
Date                                                                     
2025-06-23  201.500000  166.009995  486.000000  202.300003  168.479996   
2025-06-24  200.300003  167.740005  490.109985  203.440002  169.250000   
2025-06-25  201.559998  171.490005  492.269989  203.669998  173.360001   
2025-06-26  201.000000  174.429993  497.450012  202.639999  174.649994   
2025-06-27  201.080002  178.270004  495.940002  203.220001  178.839996   
2025-06-30  205.169998  177.389999  497.410004  207.389999  181.580002   
2025-07-01  207.820007  176.910004  492.049988  210.190002  177.220001   
2025-07-02  212.440002  179.759995  491.089996  213.339996  179.934998   
2025-07-03  213.550003  180.550003  498.839996  214.649994  180.770004   
2025-07-07  209.949997  177.559998  497.720001  216.229996  180.339996   
2025-07-08  210.009995  175.160004  496.619995  211.429993  178.789993   
2025-07-09  211.139999  177.660004  503.510010  211.330002  180.279999   
2025-07-10  212.410004  178.699997  501.480011  213.479996  179.589996   
2025-07-11  211.160004  181.309998  503.320007  212.130005  182.610001   
2025-07-14  208.619995  182.809998  503.019989  210.910004  184.729996   
2025-07-15  209.110001  183.100006  505.820007  211.889999  185.410004   
2025-07-16  210.160004  183.770004  505.619995  212.399994  185.335007   
2025-07-17  210.020004  184.699997  511.700012  211.800003  185.186996   
2025-07-18  211.179993  185.940002  510.049988  211.789993  187.360001   
2025-07-21  212.479996  191.149994  510.059998  215.779999  191.259995   

Price                          Low                                Open  \
Ticker            MSFT        AAPL        GOOG        MSFT        AAPL   
Date                                                                     
2025-06-23  487.750000  198.960007  163.330002  472.510010  201.630005   
2025-06-24  491.850006  200.199997  166.910004  486.799988  202.589996   
2025-06-25  494.559998  200.619995  168.561005  489.390015  201.449997   
2025-06-26  498.040009  199.460007  170.860001  492.809998  201.429993   
2025-06-27  499.299988  200.000000  172.710007  493.029999  201.889999   
2025-06-30  500.760010  199.259995  175.500000  495.329987  202.009995   
2025-07-01  498.049988  206.139999  174.660004  490.980011  206.669998   
2025-07-02  493.500000  208.139999  176.089996  488.700012  208.910004   
2025-07-03  500.130005  211.809998  178.190002  493.440002  212.149994   
2025-07-07  498.750000  208.800003  176.639999  495.230011  212.679993   
2025-07-08  498.200012  208.449997  173.919998  494.109985  210.100006   
2025-07-09  506.779999  207.220001  173.880005  499.739990  209.529999   
2025-07-10  504.440002  210.029999  175.740005  497.750000  210.509995   
2025-07-11  505.029999  209.860001  177.541000  497.799988  210.570007   
2025-07-14  503.970001  207.539993  180.710999  501.029999  209.929993   
2025-07-15  508.299988  208.919998  182.389999  502.790009  209.220001   
2025-07-16  506.720001  208.639999  183.080002  501.890015  210.300003   
2025-07-17  513.369995  209.589996  181.500000  505.619995  210.570007   
2025-07-18  514.640015  209.699997  184.729996  507.429993  210.869995   
2025-07-21  512.070007  211.640900  187.050003  505.589996  212.050003   

Price                                 Volume                      
Ticker            GOOG        MSFT      AAPL      GOOG      MSFT  
Date                                                              
2025-06-23  167.320007  478.209991  55814300  36975600  24864000  
2025-06-24  167.684998  488.950012  54064000  27310300  22305600  
2025-06-25  168.649994  492.040009  39525700  23627400  17495100  
2025-06-26  173.384995  492.980011  50799100  25909100  21578900  
2025-06-27  174.630005  497.549988  73188600  74053100  34539200  
2025-06-30  180.785004  497.040009  91912800  43964600  28

In [3]:
df.index.is_monotonic_increasing

True

In [14]:
df.loc[:,('Open','MSFT')]

Date
2025-06-20    482.230011
2025-06-23    478.209991
2025-06-24    488.950012
2025-06-25    492.040009
2025-06-26    492.980011
2025-06-27    497.549988
2025-06-30    497.040009
2025-07-01    496.470001
2025-07-02    489.989990
2025-07-03    493.809998
2025-07-07    497.380005
2025-07-08    497.239990
2025-07-09    500.299988
2025-07-10    503.049988
2025-07-11    498.470001
2025-07-14    501.519989
2025-07-15    503.019989
2025-07-16    505.179993
2025-07-17    505.679993
2025-07-18    514.479980
Name: (Open, MSFT), dtype: float64

In [18]:
import pandas as pd
df = pd.read_csv("https://gist.githubusercontent.com/curran/a08a1080b88344b0c8a7/raw/0e7a9b0a5d22642a06d3d5b9bcbad9890c8ee534/iris.csv")
df.head()

sepal_length  sepal_width  petal_length  petal_width species
0           5.1          3.5           1.4          0.2  setosa
1           4.9          3.0           1.4          0.2  setosa
2           4.7          3.2           1.3          0.2  setosa
3           4.6          3.1           1.5          0.2  setosa
4           5.0          3.6           1.4          0.2  setosa

## 1. 더블 인덱스 

    구글            애플
시작가 종료가   시작가 종료가

이렇게 있을 때
[시작가]만 하면 못가져 온다. 그래서 [('구글','시작가')]를 해야지 가져올 수 있다.

## 2. Read csv 했을 때 특정 index가 항상 sorting 되어있는지 확인하는 함수
 Sortindex

- P.161 ~ 167

In [19]:
import pandas as pd
import numpy as np

sepal_length  sepal_width  petal_length  petal_width    species
0             5.1          3.5           1.4          0.2     setosa
1             4.9          3.0           1.4          0.2     setosa
2             4.7          3.2           1.3          0.2     setosa
3             4.6          3.1           1.5          0.2     setosa
4             5.0          3.6           1.4          0.2     setosa
..            ...          ...           ...          ...        ...
145           6.7          3.0           5.2          2.3  virginica
146           6.3          2.5           5.0          1.9  virginica
147           6.5          3.0           5.2          2.0  virginica
148           6.2          3.4           5.4          2.3  virginica
149           5.9          3.0           5.1          1.8  virginica

[150 rows x 5 columns]

- P.161~167

In [20]:
def make_df(cols, ind):
    """빠르게 DataFrame 생성"""
    data = {c: [str(c) + str(i) for i in ind]
                for c in cols}
    return pd.DataFrame(data, ind)


# DataFrame 예제
make_df('ABC', range(3))


A   B   C
0  A0  B0  C0
1  A1  B1  C1
2  A2  B2  C2

In [ ]:
class dispaly(object):
    """Display HTML representation of multiple objects"""
    template = """<div style="float": left; padding:10px;">
     <p style='font-family:"Courier New", Courier, monospace'>{0}{1}
     """

    def __init__(self, *args):
        self.args= args
    def _repr_html_(self):
        return '\n'.join(self.template.format(a, eval(0)._repr_html_())
            for a in self.args)
    def __repr__(self):
        re


In [25]:
df.index.is_monotonic_increasing

True

In [32]:
df_state = pd.read_csv('state-abbrevs.csv')
df_state.head(), df_state.shape

(        state abbreviation
 0     Alabama           AL
 1      Alaska           AK
 2     Arizona           AZ
 3    Arkansas           AR
 4  California           CA,
 (51, 2))

In [31]:
df_area = pd.read_csv('state-areas.csv')
df_area.head(), df_area.shape

(        state  area (sq. mi)
 0     Alabama          52423
 1      Alaska         656425
 2     Arizona         114006
 3    Arkansas          53182
 4  California         163707,
 (52, 2))

In [33]:
df_population = pd.read_csv('state-population.csv')
df_population.head(), df_population.shape

(  state/region     ages  year  population
 0           AL  under18  2012   1117489.0
 1           AL    total  2012   4817528.0
 2           AL  under18  2010   1130966.0
 3           AL    total  2010   4785570.0
 4           AL  under18  2011   1125763.0,
 (2544, 4))

In [34]:
df_population.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2544 entries, 0 to 2543
Data columns (total 4 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   state/region  2544 non-null   object 
 1   ages          2544 non-null   object 
 2   year          2544 non-null   int64  
 3   population    2524 non-null   float64
dtypes: float64(1), int64(1), object(2)
memory usage: 79.6+ KB


In [ ]:
# select s.state, s.abbreviation, a.area (sq. mi), p.population
# from state s, area a, population p 
# where s.state = a.state and
#       s.abbreviation = p.state/region

In [36]:
m_df = pd.merge(df_state,df_area,how='inner',left_on='state', right_on='state')
m_df.head()

state abbreviation  area (sq. mi)
0     Alabama           AL          52423
1      Alaska           AK         656425
2     Arizona           AZ         114006
3    Arkansas           AR          53182
4  California           CA         163707

In [37]:
m_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 51 entries, 0 to 50
Data columns (total 3 columns):
 #   Column         Non-Null Count  Dtype 
---  ------         --------------  ----- 
 0   state          51 non-null     object
 1   abbreviation   51 non-null     object
 2   area (sq. mi)  51 non-null     int64 
dtypes: int64(1), object(2)
memory usage: 1.3+ KB


In [39]:
pd.merge(df_population,m_df,left_on='state/region',right_on='abbreviation')

state/region     ages  year  population    state abbreviation  \
0              AL  under18  2012   1117489.0  Alabama           AL   
1              AL    total  2012   4817528.0  Alabama           AL   
2              AL  under18  2010   1130966.0  Alabama           AL   
3              AL    total  2010   4785570.0  Alabama           AL   
4              AL  under18  2011   1125763.0  Alabama           AL   
...           ...      ...   ...         ...      ...          ...   
2443           WY  under18  1993    137458.0  Wyoming           WY   
2444           WY    total  1991    459260.0  Wyoming           WY   
2445           WY  under18  1991    136720.0  Wyoming           WY   
2446           WY  under18  1990    136078.0  Wyoming           WY   
2447           WY    total  1990    453690.0  Wyoming           WY   

      area (sq. mi)  
0             52423  
1             52423  
2             52423  
3             52423  
4             52423  
...             ...  
2443          97818  
2444          97818  
2445          97818  
2446          97818  
2447          97818  

[2448 rows x 7 columns]

In [4]:
# select s.state, s.abbreviation, a.area (sq. mi), p.population
# from  population p left out join state s on s.abbreviation = p.state/region
#   and s.abbreviation = p.state/region

In [40]:
pd.merge(df_population,m_df,how='left' ,left_on='state/region',right_on='abbreviation')

state/region     ages  year   population    state abbreviation  \
0              AL  under18  2012    1117489.0  Alabama           AL   
1              AL    total  2012    4817528.0  Alabama           AL   
2              AL  under18  2010    1130966.0  Alabama           AL   
3              AL    total  2010    4785570.0  Alabama           AL   
4              AL  under18  2011    1125763.0  Alabama           AL   
...           ...      ...   ...          ...      ...          ...   
2539          USA    total  2010  309326295.0      NaN          NaN   
2540          USA  under18  2011   73902222.0      NaN          NaN   
2541          USA    total  2011  311582564.0      NaN          NaN   
2542          USA  under18  2012   73708179.0      NaN          NaN   
2543          USA    total  2012  313873685.0      NaN          NaN   

      area (sq. mi)  
0           52423.0  
1           52423.0  
2           52423.0  
3           52423.0  
4           52423.0  
...             ...  
2539            NaN  
2540            NaN  
2541            NaN  
2542            NaN  
2543            NaN  

[2544 rows x 7 columns]

In [54]:
abbrevs = pd.read_csv('https://raw.githubusercontent.com/jakevdp/data-USstates/refs/heads/master/state-abbrevs.csv')
areas = pd.read_csv('https://raw.githubusercontent.com/jakevdp/data-USstates/refs/heads/master/state-areas.csv')
pop = pd.read_csv('https://raw.githubusercontent.com/jakevdp/data-USstates/refs/heads/master/state-population.csv')

print(abbrevs.head()); print(abbrevs.info());

        state abbreviation
0     Alabama           AL
1      Alaska           AK
2     Arizona           AZ
3    Arkansas           AR
4  California           CA
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 51 entries, 0 to 50
Data columns (total 2 columns):
 #   Column        Non-Null Count  Dtype 
---  ------        --------------  ----- 
 0   state         51 non-null     object
 1   abbreviation  51 non-null     object
dtypes: object(2)
memory usage: 944.0+ bytes
None


In [55]:
areas.head()

state  area (sq. mi)
0     Alabama          52423
1      Alaska         656425
2     Arizona         114006
3    Arkansas          53182
4  California         163707

In [56]:
pop.head()

state/region     ages  year  population
0           AL  under18  2012   1117489.0
1           AL    total  2012   4817528.0
2           AL  under18  2010   1130966.0
3           AL    total  2010   4785570.0
4           AL  under18  2011   1125763.0

In [62]:
merged = pd.merge(pop,abbrevs,how='outer',left_on='state/region', right_on='abbreviation')

merged

state/region     ages  year  population    state abbreviation
0              AK    total  1990    553290.0   Alaska           AK
1              AK  under18  1990    177502.0   Alaska           AK
2              AK    total  1992    588736.0   Alaska           AK
3              AK  under18  1991    182180.0   Alaska           AK
4              AK  under18  1992    184878.0   Alaska           AK
...           ...      ...   ...         ...      ...          ...
2539           WY  under18  1993    137458.0  Wyoming           WY
2540           WY    total  1991    459260.0  Wyoming           WY
2541           WY  under18  1991    136720.0  Wyoming           WY
2542           WY  under18  1990    136078.0  Wyoming           WY
2543           WY    total  1990    453690.0  Wyoming           WY

[2544 rows x 6 columns]

In [60]:
merged = merged.drop('abbreviation', 1) # 중복 정보 삭제
merged.head()

TypeError: drop() takes from 1 to 2 positional arguments but 3 were given

In [63]:
merged.isnull().any()

state/region    False
ages            False
year            False
population       True
state            True
abbreviation     True
dtype: bool

In [64]:
merged[merged['population'].isnull()].head()

state/region     ages  year  population state abbreviation
1872           PR  under18  1990         NaN   NaN          NaN
1873           PR    total  1990         NaN   NaN          NaN
1874           PR    total  1991         NaN   NaN          NaN
1875           PR  under18  1991         NaN   NaN          NaN
1876           PR    total  1993         NaN   NaN          NaN

In [65]:
merged.loc[merged['state/region'] == 'PR', 'state'] = 'Puerto Rico'
merged.loc[merged['state/region'] == 'USA', 'state'] = 'United States'

In [66]:
final = pd.merge(merged, areas,on='state', how='left')

In [67]:
final.isnull().any()

state/region     False
ages             False
year             False
population        True
state            False
abbreviation      True
area (sq. mi)     True
dtype: bool

In [68]:
final['state'][final['area (sq. mi)'].isnull()].unique()

array(['United States'], dtype=object)

In [76]:
data2010 = final.query("year == 2010 & ages == 'total'")
data2010.head()

state/region   ages  year  population       state abbreviation  \
43            AK  total  2010    713868.0      Alaska           AK   
51            AL  total  2010   4785570.0     Alabama           AL   
141           AR  total  2010   2922280.0    Arkansas           AR   
149           AZ  total  2010   6408790.0     Arizona           AZ   
197           CA  total  2010  37333601.0  California           CA   

     area (sq. mi)  
43        656425.0  
51         52423.0  
141        53182.0  
149       114006.0  
197       163707.0

In [77]:
data2010.set_index('state', inplace=True)
density = data2010['population'] / data2010['area (sq. mi)']

In [78]:
density.sort_values(ascending=False, inplace=True)
density.head()

state
District of Columbia    8898.897059
Puerto Rico             1058.665149
New Jersey              1009.253268
Rhode Island             681.339159
Connecticut              645.600649
dtype: float64

In [79]:
density.tail()

state
North Dakota     9.537565
Montana          6.736171
Wyoming          5.768079
Alaska           1.087509
United States         NaN
dtype: float64

In [80]:
abbrevs.head()

state abbreviation
0     Alabama           AL
1      Alaska           AK
2     Arizona           AZ
3    Arkansas           AR
4  California           CA

In [81]:
areas.head()

state  area (sq. mi)
0     Alabama          52423
1      Alaska         656425
2     Arizona         114006
3    Arkansas          53182
4  California         163707

In [82]:
pop.head()

state/region     ages  year  population
0           AL  under18  2012   1117489.0
1           AL    total  2012   4817528.0
2           AL  under18  2010   1130966.0
3           AL    total  2010   4785570.0
4           AL  under18  2011   1125763.0

In [112]:
state_region_list = pd.merge(pop,abbrevs,how='outer',left_on='state/region',right_on='abbreviation')

state_region_list = state_region_list.drop('abbreviation', axis=1)

state_region_list['state'].isnull().sum()

np.int64(96)

In [116]:
mask = state_region_list['state'].isnull()
state_region_list[mask]['state/region'].unique()

array(['PR', 'USA'], dtype=object)